### DISCUSSION TITLE
best way to extrude x-y polygons ?

### DISCUSSION THREAD

**Discussion by PinkMushroom** [marcomusy/vedo#1099] [Q&A] [answered]
I have an application where i would like to extrude polygons that lie in the x-y plane, representing traces on a printed circuit board.
What I want to implement is something very similar to the RoundedLine shape but with thickness, in the z-direction of course.

Looking how best to implement this, and experimenting with the various implementation possible methods, raised some questions

1. is there a way to extrude 2D objects directly ?
2. I looked for an extrusion function but couldn't find one. Assuming there is not a way to extrude 2D objects directly, what would be the best way to construct such an object ? It seemed to me that a collection of TetMesh objects would work, but may not be the best approach.
3.  There will be on the order of one or two thousand of these objects. Is there anything that I should know  about how to code this up efficiently ?

A final related, but maybe not directly not on-topic, question. 

I noticed that the Vedo code seems to have a very nice interface to Vtk under-the-hood. Is there a way to access that Vtk interface, aka "vtki" ?  Being able to acces various vtk classes, like vtkConvexHull2D for example, would be really, really useful !  After all, Vedo can't provide an access to ALL of Vtk, lol.

I'm just getting started but I'm finding Vedo to be really amazing !

Thank you !

**Comment by marcomusy** [comment, marked as answer]
Hi, thanks!

1. yes of course there is an extrusion [method]( https://vedo.embl.es/autodocs/content/vedo/vedo/mesh.html#Mesh.extrude), btw you can also sweep a line into a mesh with [this](https://vedo.embl.es/autodocs/content/vedo/vedo/shapes.html#Line.sweep), and stitch 2 lines with [this](https://vedo.embl.es/autodocs/content/vedo/vedo/shapes.html#Ribbon).
2. no, i would say tetmesh is not what you're looking for.
3. vedo should be able to handle that with no problem.
4. Sure! You can either access existing underlying vtk objects with eg. `mymesh.dataset` (vtkPolyData) or simply import vtk directly.

You can search for keywords in the search fields of the docs and existing  examples in the [webpage](https://vedo.embl.es/).

hope this helps

**Comment by PinkMushroom** [comment]
I was sure I put extrude in the search, but I obviously did not. In my defense I've been looking at a lot of 3D visualization toolkits so I probably got confused about which ones I had searched. lol.

Thanks very much !

**Reply by marcomusy** [reply]
No worries! Let me know if you encounter any problems.

**Comment by PinkMushroom** [comment]
here is the some code which bundles everything together that I wanted to do.  I thought it might be helpful to others as an example :-)

- generates a RoundedLine (what a great function - also the implementation is very clever :-)
- gives the line thickness using `extrude`
- then moves the lines using `LinearTransform.move`
- each trace on a different layer, i.e. with a different z-value, is given a color from the color map

I did want to get clarification on one point , just for my own education.
If I were to build these objects "by hand" you said that I should NOT use a TetMesh.
What should I use ?

```
"""Some traces generated by extruding RoundedLine shapes and then moving them in the z-direction"""

import numpy as np
import sys
from math import pi, atan, atan2, cos, sin
import vedo


# create nboxes x nboxes virtual squares which are lside length on
# each side and generate random lines within each square

nboxes = 4
lside = 1000.0
lines = []
lseg = lside/5.0
lmin = lseg/4.0
lw = 10.0

lines = []
for ii in range(nboxes):
    for jj in range(nboxes):
        points = []
        p0 = np.array((lside * ii, lside * jj))
        points = [p0]
        for kk in range(1,np.random.randint(5,9)):
            print(kk)
            # enforce a minimum length
            point = np.array((lmin+np.random.rand()*lseg,
                     lmin+np.random.rand()*lseg))
            if kk > 0:
                point = point + points[-1]
            points.append(point)
        lines.append(np.array(points))
print(lines)
        

vlines = []
for ii in range(len(lines)):
    color = vedo.colors.color_map(ii/len(lines), name="jet", vmin=0.0, vmax=1.0)
    v1 = vedo.RoundedLine(lines[ii], lw).c(color)
    v2 = v1.extrude(zshift=20.0)
    # print(v2.dataset)
    LT = vedo.LinearTransform()
    LT.translate([0.0, 0.0, ii*100.0])
    # LinearTransform move modifies the object in-place
    LT.move(v2)
    vlines.append(v2)
    
plt = vedo.Plotter(axes=11)

plt.add(vlines)
plt.show(viewup="z")

```

**Comment by marcomusy** [comment]
Hi, a TetMesh is a tetrahedral mesh e.i. a data structure which defines and describes the internal volume (typically used in FEM studies). If you don´t need that you should go for the usual polygonal surface Mesh object (eg that´s ok for 3d printing).

Also you don´t need to explicitly construct the transform, you can use the shortcut `.shift()`:

```py
import numpy as np
import vedo

# create nboxes x nboxes virtual squares which are lside length on
# each side and generate random lines within each square
nboxes = 4
lside = 1000.0
lines = []
lseg = lside / 5.0
lmin = lseg / 4.0
lw = 50.0

lines = []
for ii in range(nboxes):
    for jj in range(nboxes):
        points = []
        p0 = np.array((lside * ii, lside * jj))
        points = [p0]
        for kk in range(1, np.random.randint(5, 9)):
            # enforce a minimum length
            point = np.array(
                (lmin + np.random.rand() * lseg, lmin + np.random.rand() * lseg)
            )
            point = point + points[-1]
            points.append(point)
        lines.append(np.array(points))

vlines = []
for ii in range(len(lines)):
    color = vedo.color_map(ii / len(lines), name="jet", vmin=0, vmax=1)
    v1 = vedo.RoundedLine(lines[ii], lw).c(color)
    v2 = v1.extrude(20).shift([0.0, 0.0, ii * 100.0])
    vlines.append(v2)

plt = vedo.Plotter(axes=1)
plt.add(vlines)
plt.show()
```
<img width="519" alt="Screenshot 2024-04-22 at 10 33 31" src="https://github.com/marcomusy/vedo/assets/32848391/ef7ce80f-d277-457a-9945-624121be5d64">
